Mfem cmake

jacobmerson opened the pull request:
We are developing tools that are built using CMake and link against mfem. We have found it is much more convenient to have a version of mfem that's built with CMake as integration into the downstream project is much more streamlined.

This is an initial implementation of splitting the spack package into the `Package` and `Builder` components to support both build systems. So far, I have not tested all variants. I'm hoping that the mfem team has CI/CD for their spack builds and can add `build_system=cmake` to the mix to validate everything is working as expected.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

jacobmerson committed:
split the mfem package into the builder and package

This change is the first step in supporting both CMake and the makefile based mfem builds.

jacobmerson committed:
add CMake built to the mfem package

jacobmerson commented:
CI indicates this still isn't quite right yet.  It looks like properties such as `test_mk`, `config_mk`, `sundials_components`, `suitesparse_components` cannot be extracted from the builder? I'm not sure how to resolve this yet.

cameronrutherford commented:
@jacobmerson I took a look at the CI, and it looks like the failing builds depend on mfem. They are likely looking for the `.mk` files that are installed with the non-CMake build.

If you go into those other packages and just add `depends_on("mfem~cmake")` (or whatever your variant is), then I think CI is passing and I can also use your changes for Palace.

jacobmerson commented:
Glad this might be helpful to someone else too. I won't be able test that out for a day or two. I'm not 100 percent sure if it's that it doesn't find those files when building with cmake, it's a more complex issue with the properties being set since I switched this to the builder pattern.

cameronrutherford commented:
@jacobmerson I at least got around to concretizing and playing with these changes a little today. I found the variant `build_system=cmake` and `build_system=generic` both look available.

To rephrase my suggestion before on getting CI working, the failures are in the E4S builds of laghos and glvis. I think if you change those packages to have `depends_on("mfem build_system=generic")` that you'll get CI to pass. I think doing that at the package level and not CI level is preferred.

jacobmerson committed:
add conflicts for cmake builds for mfem downstream packages

Some packages that depend on mfem failed in CI after adding a CMakeBuilder to the package. To ensure that downstream packages will not get the wrong version of mfem, they have all been made to conflict with the CMake build.

jacobmerson commented:
I made anything that depends on mfem conflict with the CMake based builds. Hopefully this resolves the CI errors (and prevents anyone from getting unexpected behavior).

cameronrutherford commented:
Looking at the logs, it looks like the naive suggestion wasn't so lucky, and the same packages are still failing. I am still yet to try your changes locally, but I assume you're able to use this CMake build internally just fine.

Based on the `glvis` build, it looks like mfem is being built, but just isn't quite able to be consumed by other projects... https://gitlab.spack.io/spack/spack/-/jobs/16343540

Sorry I don't have any concrete suggestions to fix things, but just trying to help.

Maybe the next step to fixing this would be building either `glvis` or `laghos` locally, and getting them to play nicely with this version of the MFEM package. Not sure how involved that is, but perhaps you could get some of the maintainers to help.

jacobmerson committed:
move properties from Builder to package

jacobmerson committed:
add missing reference to package for sundials_components and suitesparse_components

jacobmerson commented:
@cameronrutherford finally got all the tests passing. I tried `laghos` with the cmake build and it surprisingly worked, so I'm not sure if the conflicts on the downstream packages could be removed. That said, it's probably safer to have the maintainers of those packages make the decision.

@tzanio @v-dobrev I have finally got all the tests passing. It would be great if you could take a look. I'm not 100% sure about the handling of the properties particularly `config_mk` and `test_mk` the other ones seem like they should be fine.

v-dobrev commented:
@jacobmerson, this sounds great, thanks for adding it! It was on my todo list but you beat me to it! 😀

I will take a look soon.

v-dobrev commented:
⚠️ @jacobmerson, just a heads up: this PR will need to be migrated to https://github.com/spack/spack-packages, e.g. using this tool: https://github.com/spack/migrate-package-prs.

jacobmerson committed:
Languages are required to make sure that spack pulls in the compiler nodes

jacobmerson commented:
@v-dobrev done. See: https://github.com/spack/spack-packages/pull/143
